src/core/algorithms/fd/eulerfd/eulerfd.cpp

```diff
@@ -0,0 +1,427 @@
+        negative_cover_.push_back(SearchTreeEulerFD(number_of_attributes_));
+    }
+
+    attribute_frequences_.resize(number_of_attributes_);
+    std::fill(attribute_frequences_.begin(), attribute_frequences_.end(), 0);
+
+    attribute_indexes_.clear();
+}
+
+void EulerFD::BuildPartition() {
+    if (number_of_tuples_ == 0) {
+        return;
+    }
+    constant_columns_ = Bitset(number_of_attributes_);
+
+    // build partition: same values in each column unite into clusters,
+    // then stripping it: clusters with 1 element cant get us new invalid fd
+    for (size_t attr_num = 0; attr_num < number_of_attributes_; attr_num++) {
+        std::unordered_map<size_t, std::vector<size_t>> values;
+        for (size_t tuple_num = 0; tuple_num < number_of_tuples_; tuple_num++) {
+            auto value = tuples_[tuple_num][attr_num];
+            auto &similar_values = values[value];
+            similar_values.push_back(tuple_num);
+        }
+        if (values.size() == 1) {
+            // its mean in this column only one value
+            constant_columns_[attr_num] = true;
+            continue;
+        } else if (values.size() == number_of_tuples_) {
+            // its mean all clusters' sizes are 1, so we can pass cycle below
+            continue;
+        } else {
+            for (auto &&[_, cluster] : values) {
+                if (cluster.size() > 1) {
+                    clusters_.emplace_back(std::move(cluster), rand_function_);
+                }
+            }
+        }
+    }
+}
+
+EulerFD::Bitset EulerFD::BuildAgreeSet(size_t t1, size_t t2) const {
+    Bitset equal_attr(number_of_attributes_);
+    for (size_t attr_num = 0; attr_num < number_of_attributes_; ++attr_num) {
+        if (tuples_[t1][attr_num] == tuples_[t2][attr_num]) {
+            equal_attr.set(attr_num);
+        }
+    }
+    return equal_attr;
+}
+
+double EulerFD::SamlingInCluster(Cluster *cluster) {
+    return cluster->Sample([this](size_t t1, size_t t2) -> size_t {
+        Bitset agree_set = BuildAgreeSet(t1, t2);
+        auto &&[_, result] = invalids_.insert(agree_set);
+
+        // check is discovered fd new
+        if (result) {
+            new_invalids_.insert(agree_set);
+            return agree_set.size() - agree_set.count();
+        } else {
+            return 0;
+        }
+    });
+}
+
+void EulerFD::Sampling() {
+    if (is_first_sample_) {
+        // in first sampling mlfq is empty, we fiil it.
+        // ww put all clusters in mlfq, even if effective was 0
+        is_first_sample_ = false;
+
+        for (size_t i = 0; i < clusters_.size(); i++) {  // TODO: maybe ranged for
+            double eff = SamlingInCluster(&clusters_[i]);
+            mlfq_.Add(&clusters_[i], eff, true);
+        }
+
+        if (mlfq_.GetLastQueueSize() > 0) {
+            effective_treshold_ = std::min(effective_treshold_, mlfq_.MaxEffectInLastQueue() / 2);
+        }
+
+        return;
+    }
+
+    // sampling in first queues of mlfq;
```

Comment:
Use the same code style for all comments and correct them using some translator. For example:
```
// Test description.
```